src/EntryPoint.sol

```diff
@@ -524,57 +286,83 @@ contract EntryPoint is EIP712, Ownable, CallContextChecker, ReentrancyGuardTrans
+    /// Currently there can only be 2 modes - simulation mode, and execution mode.
+    /// But we use a uint256 for efficient stack operations, and more flexiblity in the future.
+    /// Note: We keep the simulationFlags in the stack/memory (TSTORE doesn't work) to make sure they are reset in each new call context,
+    /// to provide protection against attacks which could spoof the execute function to believe it is in simulation mode.
+    function _execute(
+        bytes calldata encodedUserOp,
+        uint256 combinedGasOverride,
+        uint256 simulationFlags
+    ) internal virtual returns (uint256 gUsed, bytes4 err) {
         UserOp calldata u = _extractUserOp(encodedUserOp);
+
         uint256 g = Math.coalesce(uint96(combinedGasOverride), u.combinedGas);
         uint256 gStart = gasleft();
 
-        bool isSimulation = combinedGasOverride & _FLAG_IS_SIMULATION != 0;
-        unchecked {
-            // Check if there's sufficient gas left for the gas-limited self calls
-            // via the 63/64 rule. This is for gas estimation. If the total amount of gas
-            // for the whole transaction is insufficient, revert.
-            if (((gasleft() * 63) >> 6) < Math.saturatingAdd(g, _INNER_GAS_OVERHEAD)) {
-                // Don't revert if `_FLAG_IS_SIMULATION`.
-                // For `simulateExecute` to be able to get a simulation before knowing
-                // how much gas is needed without reverting.
-                if (!isSimulation) revert InsufficientGas();
+        if (
+            LibBit.or(
+                u.prePaymentAmount > u.prePaymentMaxAmount,
+                LibBit.or(
+                    u.prePaymentMaxAmount > u.totalPaymentMaxAmount,
+                    u.totalPaymentAmount > u.totalPaymentMaxAmount
+                )
+            )
+        ) {
+            err = PaymentError.selector;
+
+            if (simulationFlags == 1) {
+                revert PaymentError();
             }
-            // If `_FLAG_63_OVER_64_TEST` is set, this means `simulateExecute` just wants
-            // to check the 63/64 rule, so early return to skip the rest of the computations.
-            if (combinedGasOverride & _FLAG_63_OVER_64_TEST != 0) return (0, 0);
         }
 
+        // TODO: Why do we need this? This seems to be highly overestimating the gas required.
+        // Why do we need gExecute, if relay can check the success of the tx using the gas they provide?
+        // unchecked {
+        //     // Check if there's sufficient gas left for the gas-limited self calls
+        //     // via the 63/64 rule. This is for gas estimation. If the total amount of gas
+        //     // for the whole transaction is insufficient, revert.
+        //     if (((gasleft() * 63) >> 6) < Math.saturatingAdd(g, _INNER_GAS_OVERHEAD)) {
+        //         if(simulationFlags != 1) {
+        //             revert InsufficientGas();
+        //         }
+        //     }
+        // }
```

Comment:
The 63/64 rule was included in the original EntryPoint for anti-griefing.

Without it:
- an attacker can intentionally make the call to this EntryPoint at a very deep depth, and make the `execute` on the Delegation fail.
- an attacker can provide just enough gas to make the `execute` on the Delegation run out-of-gas, but leave enough gas for the remaining operations after the `execute` to complete.

Reply:
i think even if users trust the relayers, and even if the relayers are in a private mempool, this might still be needed.

Tbh, I'll just add it so that we don't need to fight the FUD of not being censorship resistant.

Reply:
>An attacker can intentionally make the call to this EntryPoint at a very deep depth, and make the execute on the Delegation fail (cuz the EVM has a max call depth of 1024). The idea is that with a _INNER_GAS_OVERHEAD, if the call is deep enough (about 50 depth-ish), the remaining gas will be sufficiently decayed to not pass this check.

Can we add a unit test pls for this?